pydantic/main.py

```diff
@@ -238,7 +249,46 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__custom_root_type__': _custom_root_type,
             **{n: v for n, v in namespace.items() if n not in fields},
         }
-        return super().__new__(mcs, name, bases, new_namespace, **kwargs)
+
+        cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
+
+        # Generating right signature for __init__:
+        orig_init = cls.__init__
+        orig_code = orig_init.__code__
+        # cannot modify cls.__init__ directly,
+        # as it shared between children classes, so we copy it
+        cls.__init__ = new_init = FunctionType(
+            orig_code, orig_init.__globals__, closure=orig_init.__closure__, name=orig_init.__name__
+        )
+
+        orig_posonlycount = orig_code.co_posonlyargcount if PY38 else 0
+        orig_argnames = orig_code.co_varnames[: orig_code.co_argcount + orig_code.co_kwonlyargcount + orig_posonlycount]
+        orig_argnames_set = set(orig_argnames)
+        fields_names = tuple([name for name in fields if name not in orig_argnames_set])
+        del orig_argnames_set
+
+        def fake_init():
```

Comment:
I'm not clear why we need this?

Reply:
I weren’t able to construct it from scratch. Do you have any ideas how it should be don?

Reply:
Managed to get rid of this one :)